src/wallet/migrate.cpp

```diff
@@ -0,0 +1,738 @@
+    HASH_UNSORTED = 2,   // Hash pages. Deprecated.
+    BTREE_INTERNAL = 3,  // BTree internal
+    RECNO_INTERNAL = 4,  // Recno internal
+    BTREE_LEAF = 5,      // BTree leaf
+    RECNO_LEAF = 6,      // Recno leaf
+    OVERFLOW_DATA = 7,        // Overflow
+    HASH_META = 8,       // Hash metadata
+    BTREE_META = 9,      // BTree metadata
+    QUEUE_META = 10,     // Queue Metadata
+    QUEUE_DATA = 11,     // Queue Data
+    DUPLICATE_LEAF = 12, // Off-page duplicate leaf
+    HASH_SORTED = 13,    // Sorted hash page
+
+    PAGETYPE_MAX = 14,
+};
+
+enum class RecordType : uint8_t
+{
+    KEYDATA = 1,
+    DUPLICATE = 2,
+    OVERFLOW_DATA = 3,
+    DELETE = 0x80, // Indicate this record is deleted. This is AND'd with the real type.
+};
+
+enum class BTreeFlags : uint32_t
+{
+    DUP = 1,         // Duplicates
+    RECNO = 2,       // Recno tree
+    RECNUM = 4,      // BTree: Maintain record counts
+    FIXEDLEN = 8,    // Recno: fixed length records
+    RENUMBER = 0x10, // Recno: renumber on insert/delete
+    SUBDB = 0x20,    // Subdatabases
+    DUPSORT = 0x40,  // Duplicates are sorted
+    COMPRESS = 0x80, // Compressed
+};
+
+/** Berkeley DB BTree metadata page layout */
+class MetaPage
+{
+public:
+    uint64_t lsn;             // Log Sequence Number
+    uint32_t page_num;        // Current page number
+    uint32_t magic;           // Magic number
+    uint32_t version;         // Version
+    uint32_t pagesize;        // Page size
+    uint8_t encrypt_algo;     // Encryption algorithm
+    PageType type;            // Page type
+    uint8_t metaflags;        // Meta-only flags
+    uint8_t unused1;          // Unused
+    uint32_t free_list;       // Free list page number
+    uint32_t last_page;       // Page number of last page in db
+    uint32_t partitions;      // Number of partitions
+    uint32_t key_count;       // Cached key count
+    uint32_t record_count;    // Cached record count
+    BTreeFlags flags;         // Flags
+    uint160 uid;              // Unique file ID (20 bytes, fits in uint16)
+    uint32_t unused2;         // Unused
+    uint32_t minkey;          // Minimum key
+    uint32_t re_len;          // Recno: fixed length record length
+    uint32_t re_pad;          // Recno: fixed length record pad
+    uint32_t root;            // Root page number
+    char unused3[368];        // 92 * 4 bytes of unused space
+    uint32_t crypto_magic;    // Crypto magic number
+    char trash[12];           // 3 * 4 bytes of trash space
+    unsigned char iv[20];     // Crypto IV
+    unsigned char chksum[16]; // Checksum
+
+    bool other_endian;
+    uint32_t expected_page_num;
+
+    MetaPage(uint32_t expected_page_num) : expected_page_num(expected_page_num) {}
+    MetaPage() = delete;
+
+    template<typename Stream>
+    void Unserialize(Stream& s)
+    {
+        s >> lsn;
+        s >> page_num;
+        s >> magic;
+        s >> version;
+        s >> pagesize;
+        s >> encrypt_algo;
```

Comment:
do we want to sanity check for encryption to be disabled? (didn't know bdb supported db-level encryption in the first place)

Reply:
Yes, I've added a check for that.